ir/src/lib.rs

```diff
@@ -81,47 +88,63 @@ impl AirIR {
-                    symbol_table.insert_trace_bindings(trace_bindings)?;
+                    // accumulate and save the trace bindings for later processing.
+                    trace_decls.push(trace_bindings);
                 }
                 ast::SourceSection::PublicInputs(inputs) => {
-                    // process & validate the public inputs
-                    symbol_table.insert_public_inputs(inputs)?;
                     validator.exists("public_inputs");
+                    // accumulate and save the public input bindings for later processing.
+                    pub_input_decls.extend(inputs);
                 }
                 ast::SourceSection::PeriodicColumns(columns) => {
                     // process & validate the periodic columns
                     symbol_table.insert_periodic_columns(columns)?;
                 }
                 ast::SourceSection::RandomValues(values) => {
-                    symbol_table.insert_random_values(values)?;
                     validator.exists("random_values");
+                    // save the random value declaration for later processing.
+                    rand_decls.push(values);
                 }
                 ast::SourceSection::BoundaryConstraints(stmts) => {
+                    validator.exists("boundary_constraints");
                     // save the boundary statements for processing after the SymbolTable is built.
                     boundary_stmts.extend(stmts);
-                    validator.exists("boundary_constraints");
                 }
                 ast::SourceSection::IntegrityConstraints(stmts) => {
+                    validator.exists("integrity_constraints");
                     // save the integrity statements for processing after the SymbolTable is built.
                     integrity_stmts.extend(stmts);
-                    validator.exists("integrity_constraints");
                 }
-                ast::SourceSection::EvaluatorFunction(eval_expr) => eval_exprs.push(eval_expr),
+                ast::SourceSection::EvaluatorFunction(ev_expr) => ev_decls.push(ev_expr),
             }
         }
 
         // validate sections
         validator.check()?;
 
-        // process the variable & constraint statements, and validate them against the symbol table.
+        // process evaluators
+        let mut evaluators: BTreeMap<String, Evaluator> = BTreeMap::new();
+        for ev_decl in ev_decls {
+            let constraint_builder = ConstraintBuilder::new(symbol_table.clone(), evaluators);
+            evaluators = constraint_builder.process_evaluator(ev_decl)?;
+        }
+
+        // process & validate the public inputs
+        symbol_table.insert_public_inputs(pub_input_decls)?;
 
-        // TODO: process evaluators
+        // process & validate the random value declarations
+        for rand_decls in rand_decls.into_iter() {
+            symbol_table.insert_random_values(rand_decls)?;
+        }
```

Comment:
Maybe I'm not following something, but shouldn't this happen before we process evaluators? If we don't do that, how would evaluators have access to random values?

Reply:
You're right! It was because in our discussion before we only said to process the constants and periodic columns before the evaluators, and so that was what was in the todo list, and I didn't notice the bug